File: src/main/services/session-manager.ts
import { v4 as uuid } from 'uuid'
import type { BrowserWindow } from 'electron'
import type { SessionInfo, CreateSessionRequest, PermissionResponse, PermissionRequestEvent, InteractionMode, SessionUpdateEvent, WorktreeHookProgressEvent, HookStep, ContentBlock } from '@shared/types/session'
import { applyUpdateToMessages } from '@shared/util/session-util'
import { agentManager } from './agent-manager'
import { gitService } from './git-service'
import { worktreeHookService } from './worktree-hook-service'
import { threadStore } from './thread-store'
import { workspaceService } from './workspace-service'
import { settingsService } from './settings-service'
import { logger } from '../util/logger'

/**
 * SessionManager orchestrates sessions across agent connections.
 * Each session maps to one ACP session on one agent connection.
 */
export class SessionManagerService {
  private sessions = new Map<string, SessionInfo>()
  private monitoredConnections = new Set<string>()
  private mainWindow: BrowserWindow | null = null
  private pendingPermissions = new Map<string, PermissionRequestEvent>()

  setMainWindow(window: BrowserWindow): void {
    this.mainWindow = window
  }

  /** Read enabled MCP servers from settings, mapped to the format ACP session/new expects */
  private getEnabledMcpServers(): Record<string, unknown>[] {
    const servers = settingsService.get().mcp.servers
    return servers
      .filter((s) => s.enabled)
      .map((s) => ({
        name: s.name,
        transport: s.transport,
        ...(s.command ? { command: s.command } : {}),
        ...(s.args?.length ? { args: s.args } : {}),
        ...(s.url ? { url: s.url } : {}),
        ...(s.env && Object.keys(s.env).length ? { env: s.env } : {})
      }))
  }

  private sendHookProgress(event: WorktreeHookProgressEvent): void {
    if (this.mainWindow && !this.mainWindow.isDestroyed()) {
      this.mainWindow.webContents.send('session:hook-progress', event)
    }
  }

  private ensureListener(connectionId: string): void {
    if (this.monitoredConnections.has(connectionId)) return

    const client = agentManager.getClient(connectionId)
    if (client) {
      client.on('session-update', (event: SessionUpdateEvent) => {
        const session = this.sessions.get(event.sessionId)
        if (session) {
          // Update status from session-update events
          if (event.update.type === 'status_change') {
            session.status = event.update.status
          } else if (event.update.type === 'message_complete') {
            session.status = 'active'
          } else if (event.update.type === 'error') {
            session.status = 'error'
          }
        }
      })
      client.on('permission-request', (event: PermissionRequestEvent) => {
        this.trackPermission(event)
      })
      this.monitoredConnections.add(connectionId)
    }
  }

  async createSession(request: CreateSessionRequest): Promise<SessionInfo> {
    const client = agentManager.getClient(request.connectionId)
    if (!client) {
      throw new Error(`Agent connection not found: ${request.connectionId}`)
    }

    let workingDir = request.workingDir
    let worktreePath: string | undefined
    let worktreeBranch: string | undefined
    let pendingInitialPrompt: string | undefined
    let lastHookSteps: HookStep[] | undefined
    const sessionLocalId = uuid().slice(0, 8)
    const sessionId = uuid()

    // Create git worktree if requested
    if (request.useWorktree) {
      try {
        const worktree = await gitService.createWorktree(
          request.workingDir,
          sessionLocalId
        )
        worktreePath = worktree.path
        worktreeBranch = worktree.branch
        workingDir = worktree.path
        logger.info(`Created worktree for session: ${worktreePath} (${worktreeBranch})`)
      } catch (error) {
        logger.warn('Failed to create worktree, using main directory:', error)
      }
    }

    // Execute worktree hooks (symlinks, commands, initial prompt)
    if (request.useWorktree && worktreePath) {
      try {
        pendingInitialPrompt = await worktreeHookService.executeHooks(
          request.workingDir,
          worktreePath,
          sessionId,
          (event) => {
            lastHookSteps = event.steps
            this.sendHookProgress(event)
          }
        )
      } catch (error) {
        logger.warn('Worktree hooks failed (non-fatal):', error)
      }
    }

    // Create ACP session with our stable sessionId for mapping
    const mcpServers = this.getEnabledMcpServers()
    await client.newSession(workingDir, mcpServers, sessionId)

    const session: SessionInfo = {
      sessionId,
      connectionId: request.connectionId,
      agentId: client.agentId,
      agentName: client.agentName,
      title: request.title || `Session ${sessionLocalId}`,
      createdAt: new Date().toISOString(),
      worktreePath,
      worktreeBranch,
      workingDir,
      status: 'active',
      messages: [],
      useWorktree: request.useWorktree,
      workspaceId: request.workspaceId
    }

    this.sessions.set(sessionId, session)
    this.ensureListener(request.connectionId)
    threadStore.save(session)
    logger.info(`Session created: ${sessionId} on agent ${client.agentName}`)

    // Fire-and-forget initial prompt from worktree hooks
    if (pendingInitialPrompt) {
      const promptText = pendingInitialPrompt
      const hookSteps = lastHookSteps
      setImmediate(() => {
        this.prompt(sessionId, [{ type: 'text', text: promptText }]).then(() => {
          if (hookSteps) {
            const ipStep = hookSteps.find((s) => s.label === 'Initial prompt')
            if (ipStep) ipStep.status = 'completed'
            this.sendHookProgress({ sessionId, steps: hookSteps })
          }
        }).catch((err) => {
          logger.warn('Failed to send initial prompt:', err)
          if (hookSteps) {
            const ipStep = hookSteps.find((s) => s.label === 'Initial prompt')
            if (ipStep) {
              ipStep.status = 'failed'
              ipStep.detail = String(err)
            }
            this.sendHookProgress({ sessionId, steps: hookSteps })
          }
        })
      })
    }

    return session
  }

  async forkSession(sourceSessionId: string, title?: string): Promise<SessionInfo> {
    // Find the source session
    let source = this.sessions.get(sourceSessionId)

    // Recovery: if not in memory, try to load from store
    if (!source) {
      const persisted = threadStore.loadAll().find((t) => t.sessionId === sourceSessionId)
      if (persisted) {
        source = {
          ...persisted,
          connectionId: '',
          status: 'idle'
        }
        this.sessions.set(sourceSessionId, source)
        logger.info(`Session rehydrated from store for fork: ${sourceSessionId}`)
      }
    }

    if (!source) throw new Error(`Source session not found: ${sourceSessionId}`)

    // Recovery: if connection lost, re-launch agent
    let client = agentManager.getClient(source.connectionId)
    if (!client) {
      logger.info(`Agent connection lost for source session ${sourceSessionId}, re-launching agent ${source.agentId}...`)
      const connection = await agentManager.launch(source.agentId, source.workingDir)
      source.connectionId = connection.connectionId
      this.sessions.set(sourceSessionId, source)
      client = agentManager.getClient(source.connectionId)!
      await client.newSession(source.workingDir, this.getEnabledMcpServers(), sourceSessionId)
    }

    // Verify agent connection
    if (!client) throw new Error(`Agent connection not found: ${source.connectionId}`)

    // Check agent capability
    if (!client.supportsFork) {
      throw new Error(`Agent ${client.agentName} does not support session/fork`)
    }

    // Generate IDs
    const newSessionId = uuid()

    // Call ACP fork with our stable sessionId for mapping
    await client.forkSession(sourceSessionId, source.workingDir, [], newSessionId)

    // Build the forked SessionInfo
    const session: SessionInfo = {
      sessionId: newSessionId,
      connectionId: source.connectionId,
      agentId: source.agentId,
      agentName: source.agentName,
      title: title || `Fork of ${source.title}`,
      createdAt: new Date().toISOString(),
      worktreePath: source.worktreePath,
      worktreeBranch: source.worktreeBranch,
      workingDir: source.workingDir,
      status: 'active',
      messages: [],
      useWorktree: source.useWorktree,
      workspaceId: source.workspaceId,
      parentSessionId: sourceSessionId
    }

    this.sessions.set(newSessionId, session)
    this.ensureListener(source.connectionId)
    threadStore.save(session)
    logger.info(`Session forked: ${newSessionId} from ${sourceSessionId}`)

    return session
  }

  async prompt(sessionId: string, content: ContentBlock[], mode?: InteractionMode): Promise<{ stopReason: string }> {
    let session = this.sessions.get(sessionId)
    
    // Recovery: if not in memory, try to load from store
    if (!session) {
      const persisted = threadStore.loadAll().find((t) => t.sessionId === sessionId)
      if (persisted) {
        session = {
          ...persisted,
          connectionId: '',
          status: 'idle'
        }
        this.sessions.set(sessionId, session)
        logger.info(`Session rehydrated from store: ${sessionId}`)
      }
    }

    if (!session) throw new Error(`Session not found: ${sessionId}`)

    let client = agentManager.getClient(session.connectionId)
    
    // Recovery: if connection lost, re-launch agent
    if (!client) {
      logger.info(`Agent connection lost for session ${sessionId}, re-launching agent ${session.agentId}...`)
      const connection = await agentManager.launch(session.agentId, session.workingDir)
      session.connectionId = connection.connectionId
      client = agentManager.getClient(session.connectionId)!
      
      // Re-create ACP session
      await client.newSession(session.workingDir, this.getEnabledMcpServers(), sessionId)
    }

    this.ensureListener(session.connectionId)

    // Update status
    session.status = 'prompting'

    // Add user message
    session.messages.push({
      id: uuid(),
      role: 'user',
      content: content,
      timestamp: new Date().toISOString()
    })

    // Subscribe directly to session-update events for this prompt.
    // This ensures agent messages are captured in session.messages
    // before persistence.
    const promptListener = (event: SessionUpdateEvent): void => {
      if (event.sessionId === sessionId) {
        session.messages = applyUpdateToMessages(session.messages, event.update)
      }
    }
    client.on('session-update', promptListener)

    try {
      const result = await client.prompt(sessionId, content, mode)

      session.status = 'active'
      // Persist messages after prompt completes
      threadStore.updateMessages(sessionId, session.messages)
      return { stopReason: result.stopReason }
    } catch (error) {
      session.status = 'error'
      // Still persist messages on error so conversation history is saved
      threadStore.updateMessages(sessionId, session.messages)
      throw error
    } finally {
      client.removeListener('session-update', promptListener)
    }
  }

  async cancel(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)

    const client = agentManager.getClient(session.connectionId)
    if (client) {
      client.cancel(sessionId)
    }

    session.status = 'cancelled'
  }

  resolvePermission(response: PermissionResponse): void {
    // Remove from pending tracking
    this.pendingPermissions.delete(response.requestId)

    // Forward the permission response to all active agent connections.
    // Each client will check if it has a pending resolver for this requestId.
    const connections = agentManager.listConnections()
    for (const conn of connections) {
      const client = agentManager.getClient(conn.connectionId)
      if (client) {
        client.resolvePermission(response)
      }
    }
  }

  trackPermission(event: PermissionRequestEvent): void {
    this.pendingPermissions.set(event.requestId, event)
  }

  listPendingPermissions(): PermissionRequestEvent[] {
    return Array.from(this.pendingPermissions.values())
  }

  async setMode(sessionId: string, modeId: string): Promise<void> {
    const session = this.sessions.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)
    const client = agentManager.getClient(session.connectionId)
    if (!client) throw new Error(`Agent connection not found: ${session.connectionId}`)
    await client.setMode(sessionId, modeId)
  }

  async setConfigOption(sessionId: string, configId: string, value: string): Promise<unknown> {
    const session = this.sessions.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)
    const client = agentManager.getClient(session.connectionId)
    if (!client) throw new Error(`Agent connection not found: ${session.connectionId}`)
    return await client.setConfigOption(sessionId, configId, value)
  }

  /**
   * Auto-generate a thread title using the configured summarization agent.
   * Launches the agent, sends the conversation as context, and extracts a short title.
   */
  async generateTitle(sessionId: string): Promise<string | null> {
    const settings = settingsService.get()
    const agentId = settings.general.summarizationAgentId
    if (!agentId) {
      logger.warn('generateTitle: no summarization agent configured')
      return null
    }

    let session = this.sessions.get(sessionId)
    // Also check persisted threads if not in memory
    if (!session) {
      const persisted = threadStore.loadAll().find((t) => t.sessionId === sessionId)
      if (persisted) {
        session = {
          ...persisted,
          connectionId: '',
          agentId: '',
          status: 'idle',
          messages: persisted.messages.map((m) => ({
            ...m,
            content: m.content as ContentBlock[]
          }))
        }
        logger.info(`generateTitle: loaded session from persisted store, workingDir: ${session.workingDir}`)
      }
    }
    if (!session) {
      logger.warn(`generateTitle: session not found: ${sessionId}`)
      return null
    }

    // Build a summary of the conversation for the title prompt
    const conversationText = session.messages
      .filter((m) => m.content.length > 0)
      .map((m) => {
        const text = m.content
          .filter((b): b is { type: 'text'; text: string } => b.type === 'text')
          .map((b) => b.text)
          .join('\n')
        return `${m.role === 'user' ? 'User' : 'Agent'}: ${text}`
      })
      .join('\n\n')

    if (!conversationText.trim()) {
      logger.warn(`generateTitle: no conversation text for session: ${sessionId}, message count: ${session.messages.length}`)
      return null
    }

    logger.info(`generateTitle: generating title for session ${sessionId} with ${session.messages.length} messages`)

    const titlePrompt = `Generate a very short title (max 6 words) for the following conversation. Reply with ONLY the title, nothing else. No quotes, no punctuation at the end.\n\n${conversationText}`

    try {
      // Find or launch the summarization agent
      const connections = agentManager.listConnections()
      let connection = connections.find(
        (c) => c.agentId === agentId && c.status === 'connected'
      )
      if (!connection) {
        connection = await agentManager.launch(agentId, session.workingDir)
      }

      const client = agentManager.getClient(connection.connectionId)
      if (!client) return null

      // Create a temporary session for the title generation
      const tempSessionId = `title-${uuid().slice(0, 8)}`
      await client.newSession(session.workingDir, this.getEnabledMcpServers(), tempSessionId)

      // Collect response text from streaming events
      let responseText = ''
      const listener = (event: SessionUpdateEvent): void => {
        if (event.sessionId !== tempSessionId) return
        if (event.update.type === 'text_chunk' && event.update.text) {
          responseText += event.update.text
        }
      }
      client.on('session-update', listener)

      try {
        await client.prompt(tempSessionId, titlePrompt)
      } finally {
        client.removeListener('session-update', listener)
      }

      // Clean up: we don't persist the temp session
      const title = responseText.trim().replace(/^["']|["']$/g, '').slice(0, 100)
      if (!title) return null

      // Apply the generated title
      threadStore.rename(sessionId, title)
      session.title = title

      // Notify the renderer
      if (this.mainWindow && !this.mainWindow.isDestroyed()) {
        this.mainWindow.webContents.send('session:update', {
          sessionId,
          update: { type: 'session_info_update', title }
        })
      }

      logger.info(`Auto-generated title for ${sessionId}: ${title}`)
      return title
    } catch (error) {
      logger.warn(`Failed to auto-generate title for ${sessionId}:`, error)
      return null
    }
  }

  getSession(sessionId: string): SessionInfo | undefined {
    return this.sessions.get(sessionId)
  }

  listSessions(): SessionInfo[] {
    return Array.from(this.sessions.values())
  }

  async removeSession(sessionId: string, cleanupWorktree = false): Promise<void> {
    const session = this.sessions.get(sessionId)
    // Also check persisted threads if not in memory
    const persisted = !session ? threadStore.loadAll().find((t) => t.sessionId === sessionId) : null
    const thread = session || persisted

    if (!thread) return

    // Terminate the agent connection first so it releases file handles on the worktree
    const connectionId = session?.connectionId
    if (connectionId) {
      try {
        agentManager.terminate(connectionId)
        logger.info(`Agent connection terminated for session: ${sessionId}`)
        // Give the OS a moment to release file handles (Windows needs this)
        if (process.platform === 'win32') {
          await new Promise((resolve) => setTimeout(resolve, 500))
        }
      } catch (error) {
        logger.warn('Failed to terminate agent connection:', error)
      }
    }

    // Clean up worktree if requested
    if (cleanupWorktree && thread.worktreePath && thread.useWorktree) {
      try {
        // For worktree removal we need the original project path (workspace path, not the worktree itself)
        const workspaces = workspaceService.list()
        const workspace = workspaces.find((w) => w.id === thread.workspaceId)
        if (workspace) {
          await gitService.removeWorktree(workspace.path, thread.worktreePath)
          logger.info(`Worktree removed: ${thread.worktreePath}`)
        }
      } catch (error) {
        logger.warn('Failed to clean up worktree:', error)
      }
    }

    // Orphan children: promote child sessions to root level
    for (const [, s] of this.sessions) {
      if (s.parentSessionId === sessionId) {
        s.parentSessionId = undefined
        threadStore.save(s)
      }
    }

    this.sessions.delete(sessionId)
    threadStore.remove(sessionId)
  }
}

export const sessionManager = new SessionManagerService()

File: src-tauri/src/services/session_manager.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value as JsonValue};
use uuid::Uuid;
use log::{info, warn};
use tauri::AppHandle;
use chrono::Utc;

use super::agent_manager::AgentManager;
use super::settings_service::SettingsService;
use super::thread_store::ThreadStore;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionInfo {
    pub session_id: String,
    pub connection_id: String,
    pub agent_id: String,
    pub agent_name: String,
    pub title: String,
    pub created_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub worktree_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub worktree_branch: Option<String>,
    pub working_dir: String,
    pub status: String,
    pub messages: Vec<JsonValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interaction_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub use_worktree: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionRequest {
    pub connection_id: String,
    pub working_dir: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub use_worktree: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interaction_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub branch_name: Option<String>,
}

pub struct SessionManager {
    sessions: HashMap<String, SessionInfo>,
    pending_permissions: HashMap<String, JsonValue>,
}

impl SessionManager {
    pub fn new() -> Self {
        Self {
            sessions: HashMap::new(),
            pending_permissions: HashMap::new(),
        }
    }

    pub async fn create_session(
        &mut self,
        request: CreateSessionRequest,
        agents: &mut AgentManager,
        settings: &SettingsService,
        thread_store: &ThreadStore,
        git_worktree_path: Option<String>,
        git_worktree_branch: Option<String>,
        app_handle: &AppHandle,
    ) -> Result<SessionInfo, String> {
        let client = agents.get_client_mut(&request.connection_id)
            .ok_or_else(|| format!("Agent connection not found: {}", request.connection_id))?;

        let session_id = Uuid::new_v4().to_string();
        let session_local_id = &session_id[..8];

        let working_dir = git_worktree_path.as_deref()
            .unwrap_or(&request.working_dir)
            .to_string();

        // Get enabled MCP servers
        let mcp_servers = get_enabled_mcp_servers(settings);

        // Create ACP session
        client.new_session(
            &working_dir,
            mcp_servers,
            Some(&session_id),
            request.interaction_mode.as_deref(),
            app_handle,
        ).await?;

        // Set mode/model
        if let Some(ref mode) = request.interaction_mode {
            if let Err(e) = client.set_mode(&session_id, mode).await {
                warn!("Failed to set mode: {}", e);
            }
        }
        if let Some(ref model) = request.model_id {
            if let Err(e) = client.set_model(&session_id, model).await {
                warn!("Failed to set model: {}", e);
            }
        }

        let session = SessionInfo {
            session_id: session_id.clone(),
            connection_id: request.connection_id.clone(),
            agent_id: client.agent_id.clone(),
            agent_name: client.agent_name.clone(),
            title: request.title.unwrap_or_else(|| format!("Session {}", session_local_id)),
            created_at: Utc::now().to_rfc3339(),
            worktree_path: git_worktree_path,
            worktree_branch: git_worktree_branch,
            working_dir,
            status: "active".to_string(),
            messages: vec![],
            interaction_mode: request.interaction_mode,
            use_worktree: request.use_worktree,
            workspace_id: request.workspace_id,
            parent_session_id: None,
            branch_name: request.branch_name,
        };

        // Persist
        let session_value = serde_json::to_value(&session).unwrap_or_default();
        let _ = thread_store.save(&session_value);

        self.sessions.insert(session_id.clone(), session.clone());
        info!("Session created: {} on {}", session_id, client.agent_name);

        Ok(session)
    }

    pub async fn prompt(
        &mut self,
        session_id: &str,
        content: JsonValue,
        mode: Option<String>,
        agents: &mut AgentManager,
        settings: &SettingsService,
        thread_store: &ThreadStore,
        app_handle: &AppHandle,
    ) -> Result<String, String> {
        let session = self.sessions.get_mut(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        let connection_id = session.connection_id.clone();
        let working_dir = session.working_dir.clone();

        session.status = "prompting".to_string();

        // Add user message
        let user_msg = json!({
            "id": Uuid::new_v4().to_string(),
            "role": "user",
            "content": content,
            "timestamp": Utc::now().to_rfc3339()
        });
        session.messages.push(user_msg);

        if let Some(ref m) = mode {
            session.interaction_mode = Some(m.clone());
        }

        let client = agents.get_client_mut(&connection_id)
            .ok_or_else(|| format!("Agent connection lost for session: {}", session_id))?;

        let result = client.prompt(session_id, content, mode.as_deref()).await;

        let session = self.sessions.get_mut(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        match &result {
            Ok(_) => session.status = "active".to_string(),
            Err(_) => session.status = "error".to_string(),
        }

        // Persist messages
        let messages = session.messages.clone();
        let _ = thread_store.update_messages(
            session_id,
            &session.working_dir,
            &json!(messages),
        );

        result
    }

    pub fn cancel(&mut self, session_id: &str, agents: &AgentManager) -> Result<(), String> {
        let session = self.sessions.get(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        // Cancel any pending permissions for this session
        let to_cancel: Vec<String> = self.pending_permissions.iter()
            .filter(|(_, p)| p["sessionId"].as_str() == Some(session_id))
            .map(|(id, _)| id.clone())
            .collect();

        for req_id in to_cancel {
            if let Some(conn) = agents.connections.get(&session.connection_id) {
                conn.resolve_permission(&req_id, "__cancelled__");
            }
            self.pending_permissions.remove(&req_id);
        }

        // Cancel via ACP
        if let Some(client) = agents.connections.get(&session.connection_id) {
            client.cancel(session_id);
        }

        if let Some(s) = self.sessions.get_mut(session_id) {
            s.status = "cancelled".to_string();
        }

        Ok(())
    }

    pub fn resolve_permission(
        &mut self,
        request_id: &str,
        option_id: &str,
        agents: &AgentManager,
        app_handle: &AppHandle,
    ) {
        self.pending_permissions.remove(request_id);

        let _ = app_handle.emit("session:permission-resolved", json!({
            "requestId": request_id
        }));

        // Forward to all agent connections
        for client in agents.connections.values() {
            client.resolve_permission(request_id, option_id);
        }
    }

    pub fn get_session(&self, session_id: &str) -> Option<&SessionInfo> {
        self.sessions.get(session_id)
    }

    pub fn list_sessions(&self) -> Vec<SessionInfo> {
        self.sessions.values().cloned().collect()
    }

    pub fn rename(&mut self, session_id: &str, title: &str, thread_store: &ThreadStore) {
        if let Some(session) = self.sessions.get_mut(session_id) {
            session.title = title.to_string();
            let _ = thread_store.rename(session_id, &session.working_dir, title);
        }
    }

    pub fn remove_session(
        &mut self,
        session_id: &str,
        agents: &mut AgentManager,
        thread_store: &ThreadStore,
    ) {
        if let Some(session) = self.sessions.remove(session_id) {
            agents.terminate(&session.connection_id);
            let _ = thread_store.remove(session_id, &session.working_dir);
        }
    }

    pub async fn set_mode(
        &mut self,
        session_id: &str,
        mode_id: &str,
        agents: &mut AgentManager,
    ) -> Result<(), String> {
        let conn_id = self.sessions.get(session_id)
            .map(|s| s.connection_id.clone())
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        let client = agents.get_client_mut(&conn_id)
            .ok_or_else(|| format!("Connection not found: {}", conn_id))?;
        client.set_mode(session_id, mode_id).await
    }

    pub async fn set_model(
        &mut self,
        session_id: &str,
        model_id: &str,
        agents: &mut AgentManager,
    ) -> Result<(), String> {
        let conn_id = self.sessions.get(session_id)
            .map(|s| s.connection_id.clone())
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        let client = agents.get_client_mut(&conn_id)
            .ok_or_else(|| format!("Connection not found: {}", conn_id))?;
        client.set_model(session_id, model_id).await
    }

    pub async fn set_config_option(
        &mut self,
        session_id: &str,
        config_id: &str,
        value: &str,
        agents: &mut AgentManager,
    ) -> Result<JsonValue, String> {
        let conn_id = self.sessions.get(session_id)
            .map(|s| s.connection_id.clone())
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        let client = agents.get_client_mut(&conn_id)
            .ok_or_else(|| format!("Connection not found: {}", conn_id))?;
        client.set_config_option(session_id, config_id, value).await
    }
}

fn get_enabled_mcp_servers(settings: &SettingsService) -> Vec<JsonValue> {
    settings.get().mcp.servers.iter()
        .filter(|s| s.enabled)
        .map(|s| {
            let mut obj = json!({
                "name": s.name,
                "transport": s.transport
            });
            if let Some(ref cmd) = s.command { obj["command"] = json!(cmd); }
            if let Some(ref args) = s.args { if !args.is_empty() { obj["args"] = json!(args); } }
            if let Some(ref url) = s.url { obj["url"] = json!(url); }
            if let Some(ref env) = s.env { if !env.is_empty() { obj["env"] = json!(env); } }
            obj
        })
        .collect()
}
